File: Programming Language/assignment_1/primeNumber.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <math.h>
#include <time.h>
#define N 30
#define M 100000
bool p[M+1];
int num[N];
int GCD(int, int);
void mergeSort(int, int);
void merge(int, int, int);
void primeNumber(int);
int prime(int, int);
int main(void)
{
	int i, n, tmp, gcd, max=0;
	clock_t start, end;

	while(1){
		printf(">> Input the number of numbers to process: ");
		scanf("%d", &n);

		if(2<=n && n<=30)
			break;
	}

	printf(">> Input the numbers to be processed:\n");
	for(i=0; i<n; i++){
		scanf("%d", &tmp);

		if(1<=tmp && tmp<=100000){
			num[i] = tmp;

			if(tmp>max)
				max = tmp;
		}
		else{
			i--;
			printf("Error: 1<=N<=100,000\n");
			continue;
		}
	}

	start = clock();

	gcd = GCD(num[0], num[1]);
	for(i=2; i<n; i++)
		gcd = GCD(gcd, num[i]);

	printf("GCD of input numbers is %d\n", gcd);
	
	mergeSort(0, n-1);
	memset(p, true, sizeof(p));
	primeNumber(max);

	for(i=0; i+1<n; i++){
		if(num[i]!=num[i+1])
	   		printf("Number of prime numbers between %d, %d: %d\n", num[i], num[i+1], prime(num[i], num[i+1]));
	}

	end = clock();

	printf("Total execution time using C is %lf seconds!\n", (double)(end-start)/CLOCKS_PER_SEC);

	return 0;
}
int GCD(int a, int b)
{
	int i;

	if(a>b) i=b;
	else i=a;

	while(i>0)
	{
		if((a%i==0)&&(b%i==0)) break;
		i--;
	}

	return i;
}
void merge(int start, int mid, int end)
{
	int tmp[N];
	int i, j, k;

	i=start; j=mid+1; k=0;

	while(i<=mid && j<=end){
		if(num[i]<=num[j])
			tmp[k++] = num[i++];
		else
			tmp[k++] = num[j++];
	}

	while(i<=mid)
		tmp[k++] = num[i++];

	while(j<=end)
		tmp[k++] = num[j++];

	k--;

	while(k>=0)
	{
		num[start+k] = tmp[k];
		k--;
	}

}
void mergeSort(int start, int end)
{
	int mid;

	if(start<end){
		mid = (start + end)/2;
		mergeSort(start, mid);
		mergeSort(mid+1, end);
		merge(start, mid, end);
	}
}
void primeNumber(int m)
{
	p[0] = false;
	p[1] = false;

	for(int i=2; i<=sqrt(m); i++){
		if(p[i]){
			for(int j=i*2; j<=m; j+=i)
				p[j] = false;
		}
	}
}
int prime(int a, int b)
{
	int num=0;

	while(a<=b){
		if(p[a])
			num++;

		a++;
	}

	return num;
}

File: Programming Language/assignment_1/primeNumber.java
import java.util.Scanner;
import java.util.Arrays;

public class primeNumber {

    static int number[] = new int[30];
    static boolean p[] = new boolean[100001];

    public static int GCD(int a, int b)
    {
        int i;

        if(a > b) i = b;
        else i = a;

        while(i>0){
            if((a%i==0)&&(b%i==0)) break;
            i--;
        }

        return i;
    }

    private static void merge(int start, int mid, int end)
    {
        int tmp[] = new int[30];
        int i, j, k;

        i=start; j=mid+1; k=0;

        while(i<=mid && j<=end){
            if(number[i]<=number[j])
                tmp[k++] = number[i++];
            else
                tmp[k++] = number[j++];
        }

        while(i<=mid)
            tmp[k++] = number[i++];

        while(j<=mid)
            tmp[k++] = number[j++];

        k--;

        while(k>=0){
            number[start+k] = tmp[k];
            k--;
        }
    }

    private static void mergeSort(int start, int end)
    {
        int mid;

        if(start<end){
            mid = (start + end)/2;
            mergeSort(start, mid);
            mergeSort(mid+1, end);
            merge(start, mid, end);
        }
    }

    private static void primeNumber(int m)
    {
        p[0] = false;
        p[1] = false;

        for (int i = 2; i <= Math.sqrt(m); i++) {
            if (p[i]) {
                for (int j = i * 2; j <= m; j += i)
                    p[j] = false;
            }
        }
    }

    private static int prime(int a, int b)
    {
        int num = 0;

        while(a<=b){
            if(p[a])
                num++;

            a++;
        }

        return num;
    }

    public static void main(String[] args) {

        Scanner scanner = new Scanner(System.in);
        int gcd, tmp, max=0, i, n;

        while(true) {
            System.out.print(">> Input the number of numbers to process: ");
            n = scanner.nextInt();

            if(2<=n && n<=30)
                break;
        }

        System.out.println(">> Input the numbers to be processed:");
        for(i=0; i<n; i++) {
            tmp = scanner.nextInt();

            if(1<=tmp && tmp<=100000) {
                number[i] = tmp;

                if (tmp > max)
                    max = tmp;
            }
            else{
                i--;
                System.out.print("Error: 1<=N<=100,000\n");
                continue;
            }
        }

        long start = System.nanoTime();

        gcd = GCD(number[0], number[1]);
        for(i=2; i<n; i++)
            gcd = GCD(gcd, number[i]);

        System.out.println("GCD of input numbers is " + gcd);

        mergeSort(0, n-1);
        Arrays.fill(p, true);
        primeNumber(max);

        for(i=0; i+1<n; i++){
            if(number[i]!=number[i+1])
                System.out.println("Number of prime numbers between "
                        + number[i] + ", " + number[i+1] + ": " + prime(number[i], number[i+1]));
        }

        long end = System.nanoTime();

        System.out.println("Total execution time using JAVA is " + (double)(end-start)/1000000000 + " seconds!");
    }
}